src/harness/vpath.ts

```diff
@@ -0,0 +1,437 @@
+
+        // Validate component strings
+        if (invalidRootComponentRegExp.test(components[0])) return false;
+        for (let i = 1; i < components.length; i++) {
+            if (invalidComponentRegExp.test(components[i])) return false;
+        }
+
+        return true;
+    }
+
+    const absolutePathRegExp = /^[\\/]([\\/](.*?[\\/](.*?[\\/])?)?)?|^[a-zA-Z]:[\\/]?|^\w+:\/{2}[^\\/]*\/?/;
+
+    function getRootLength(path: string) {
+        const match = absolutePathRegExp.exec(path);
+        return match ? match[0].length : 0;
+    }
+
+    /**
+     * Determines whether a path is an absolute path (e.g. starts with `/`, `\\`, or a dos path
+     * like `c:`).
+     */
+    export function isAbsolute(path: string) {
+        return absolutePathRegExp.test(path);
+    }
+
+    /**
+     * Determines whether a path consists only of a path root.
+     */
+    export function isRoot(path: string) {
+        const rootLength = getRootLength(path);
+        return rootLength > 0 && rootLength === path.length;
+    }
+
+    const trailingSeperatorRegExp = /[\\/]$/;
+
+    /**
+     * Determines whether a path has a trailing separator (`/`).
+     */
+    export function hasTrailingSeparator(path: string) {
+        return trailingSeperatorRegExp.test(path) && !isRoot(path);
+    }
+
+    /**
+     * Adds a trailing separator (`/`) to a path if it doesn't have one.
+     */
+    export function addTrailingSeparator(path: string) {
+        return !trailingSeperatorRegExp.test(path) && path ? path + "/" : path;
+    }
+
+    /**
+     * Removes a trailing separator (`/`) from a path if it has one.
+     */
+    export function removeTrailingSeparator(path: string) {
+        return trailingSeperatorRegExp.test(path) && !isRoot(path) ? path.slice(0, -1) : path;
+    }
+
+    function reduce(components: ReadonlyArray<string>) {
+        const normalized = [components[0]];
+        for (let i = 1; i < components.length; i++) {
+            const component = components[i];
+            if (component === ".") continue;
+            if (component === "..") {
+                if (normalized.length > 1) {
+                    if (normalized[normalized.length - 1] !== "..") {
+                        normalized.pop();
+                        continue;
+                    }
+                }
+                else if (normalized[0]) continue;
+            }
+            normalized.push(component);
+        }
+        return normalized;
+    }
+
+    /**
+     * Normalize a path containing path traversal components (`.` or `..`).
+     */
+    export function normalize(path: string): string {
+        const components = reduce(parse(path));
+        return components.length > 1 && hasTrailingSeparator(path) ? format(components) + sep : format(components);
+    }
+
+    /**
+     * Combines two or more paths. If a path is absolute, it replaces any previous path.
+     */
+    export function combine(path: string, ...paths: string[]) {

```

Comment:
many of the helpers here have counterparts in core, can we reuse?

Reply:
Yes, where possible. I like the logical grouping of path-related functionality in the `vpath` namespace (along with the more concise naming), and wonder if it would make sense to similarly consolidate related functionality in the compiler to reduce the volume of functions that show up in completions.

Reply:
that is fine, but as we chatted offline, i think this restructuring should be done in the core, and not in the tests, and not by introducing a set of duplicate implementations.

Reply:
I've removed most of the strict duplications already. What remains are places where the implementations are divergent. When you say "core", are you speaking of a new `src/core` folder for those shared dependencies or refactoring existing functions in `src/compiler/core.ts` to support the needs of both compiler and harness? I haven't made this change yet as I didn't want to bloat compiler with harness-specific functionality unnecessarily.